Introduce a variety of new ways for embedding stories

swissspidy opened the pull request:
## Context

In order to expand the capabilities of the plugin for developers and site owners, we've teamed up with **rtCamp** to design and implement a range of new features and enhancements that brings multiple new ways for embedding stories on their website.

This PR is the result of that work.

**Kudos to everyone involved!** 🎉 🥳 💪 🚀 👏 

## Summary

[See this WIP doc for more details](https://docs.google.com/document/d/1yPg0yhIKomzPr_5UJ4h28RnSyd8AgD6isRVmtUef-nE/edit#)

### Gutenberg Block

Web Stories for WordPress has previously offered a block to embed a single web story from any provided URL. Now, this has been extended to support displaying multiple stories with a variety of layout options and overall improved UX.

Highlights:

1. Choose between embedding a single story, a set of selected stories, or your latest stories
2. Stories can be displayed in a list, a grid, a carousel, or a carousel with circles
3. Stories are displayed with a nice placeholder image and can be opened in a lightbox

### Customizer Integration

Themes can opt in to support web stories using `add_theme_support()`, which allows them to display stories in a location of their choice. This will unlock a new section in the customizer to control how the stories should be displayed. 

While this is only available if the user’s current theme has explicitly declared support for Web Stories, the plugin itself adds such support for all WordPress default themes (e.g. Twenty Twenty, Twenty Twenty-One), which are very popular, to enable this experience for many users out of the box.

### Shortcodes

The existing `[web_stories_embed]` shortcode allows embedding a single web story by its URL.

Now there's a new accompanying `[web_stories]` shortcode that allows displaying a list of recent stories from the site.

### Classic Editor

Some sites are still using the classic editor instead of the block editor. These users had to previously use a (not properly documented) shortcode to display stories.

Now there’s a direct integration with the TiinyMCE-based editor that allows them to configure and insert this new shortcode.

### Sidebar Widget

Widgets are another way for users to display content across their site, beyond the post content area. While they will be replaced by blocks eventually, many sites and themes are still relying a lot on widgets.

The Web Stories plugin now includes a widget to display the latest stories on the site in places like the sidebar or any other widget area.

## Relevant Technical Choices

<!-- Please describe your changes. -->

## To-do

<!-- A list of things that need to be addressed in this PR or follow-up changes. -->

## User-facing changes

### Gutenberg Block

https://user-images.githubusercontent.com/841956/110138629-771e2300-7dd2-11eb-8be7-c111e82db2b8.mov

### Customizer Integration

https://user-images.githubusercontent.com/841956/110117268-ebe46380-7db8-11eb-916b-46a85c45c179.mov

### Classic Editor

https://user-images.githubusercontent.com/841956/110117834-b8560900-7db9-11eb-9097-65cb14e64a4f.mov

### Sidebar Widget

https://user-images.githubusercontent.com/841956/110122407-e1799800-7dbf-11eb-921d-490eb190edb5.mov


## Testing Instructions

<!--
How can the changes in this PR be verified?
Please provide step-by-step instructions how to reproduce the issue, if applicable.
Write step-by-step test instructions aimed at non-tech-savvy users, even if the PR is not user-facing.
-->

### QA

<!--
Not all changes require manual QA.
-->

<!-- ignore-task-list-start -->
- [ ] This is a non-user-facing change and requires no QA
<!-- ignore-task-list-end -->

This PR can be tested by following these steps:

#### Gutenberg Block

**Block Migration**

1. Using version 1.4.0 or `main` of the plugin, insert a Web Stories block in a new blog post to embed a single story
2. Check out the version from this PR
3. Edit the blog post and verify that the story embed still works as expected

**General Usage**

1. Insert a Web Stories block in a new blog post
2. Test the various layout and filtering options

#### Customizer Integration

**Core Themes**

1. Install all the WordPress core default themes (from Twenty Ten to Twenty Twenty-One)
2. For each default theme, verify that there's a new "Web Stories" section in the Customizer
3. In that section, enable to show stories and use the various options to display stories
3. Verify that only the "Circles" layout can be selected

**Other Theme**

1. Install a non-default WordPress theme
2. Verify that "Web Stories" section is not present in the Customizer

#### Classic Editor TinyMCE Button

1. Create a new post using the classic editor
2. Verify that there's a new Web Stories button in the TinyMCE toolbar that allows inserting stories via the new `[web_stories]` shortcode
3. Verify that all the available options work as expected
4. Verify that the presentation on the frontend matches what has been configured

#### Sidebar Widget

1. Insert the new Web Stories widget somewhere
2. Verify that all the available options work as expected

### UAT

<!--
Sometimes the testing instructions for UAT can differ from the ones for QA.
-->

<!-- ignore-task-list-start -->
- [ ] UAT should use the same steps as above.
<!-- ignore-task-list-end -->

<!--
If the above checkbox has not been checked, write down all steps necessary for user acceptance testing take to test this PR.
-->
This PR can be tested by following these steps:

1.

## Reviews

### Does this PR have a security-related impact?

<!-- Examples: new APIs, changes to KSES, etc.  -->

Noe

### Does this PR change what data or activity we track or use?

<!-- Examples: changes to telemetry, new third-party APIs -->

No

### Does this PR have a legal-related impact?

<!-- Examples: new images with unknown sources, new production dependencies with incompatible licenses -->

No

## Checklist

<!-- Check these after PR creation -->

- [X] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [X] I have tested this code to the best of my abilities
- [X] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testiing.md))
- [X] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [X] This PR contains automated tests (unit, integration, and/or e2e) to verify the code works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [X] I have added documentation where necessary
- [X] I have added a matching `Type: XYZ` label to the PR

---

<!--
Please reference the issue(s) this PR addresses.
No URLs, just the issue numbers.
Use "Fixes #123" if it fixes an issue.
Example:

Fixes #123
Partially addresses #456
See #789
-->

Fixes #6078
See #4863

Also fixes #6511

divyarajmasani committed:
- Feature: Adds Web Stories Block.
- Feature: Adds Block type and view type selection panel.

Co-authored-by: Rahi Prajapati <rahi.prajapati@rtcamp.com>

divyarajmasani committed:
- Feature: Adds Web Stories Block.
- Feature: Adds Block type and view type selection panel.

Co-authored-by: Rahi Prajapati <rahi.prajapati@rtcamp.com>

divyarajmasani committed:
- Feature: Adds selected stories block.
- Change: Renames Embed story input prop name.

Co-authored-by: Edi Amin <edi.amin@rtcamp.com>

Commit:
Stories Widget

Commit:
Fix: Tests

Commit:
Fix: PHPCS

Commit:
Method return type

Commit:
Methods return type

Commit:
Fix: phpcs

Commit:
Fix: Syntax

Commit:
Tests: Stories widget class

Commit:
Use assertEqualSets for php 5.6

Commit:
Web stories button for TinyMCE Editor

divyarajmasani committed:
Feature: Adds lightbox support for the web stories block, stories now open in a lightbox.

Co-authored-by: Rahi Prajapati <rahi.prajapati@rtcamp.com>

divyarajmasani committed:
Feature: Uses 'glider-js' for the carousel view.
Feature: Circles view is displayed as carousel.

divyarajmasani committed:
- Updates blockType and viewType icons for the 'Web Stories List' block.

Commit:
Separate functions for layouts and orders

Commit:
Assign variable for view type

Commit:
Label position

Commit:
Jest Tests

Commit:
Snapshot tests

Commit:
Unit tests for store modules

Commit:
Store selector tests

Commit:
Add carousel icon

Commit:
Web stories icon for MCE button

divyarajmasani committed:
- Uses CSS custom properties for some styling aspects for easy customization
- Adds base styles for lightbox and carousel view type.
- Adds styles for the stories block, uses pre-processor for the styles.

divyarajmasani committed:
- Uses CSS custom properties for some styling aspects for easy customization
- Adds base styles for lightbox and carousel view type.
- Adds styles for the stories block, uses pre-processor for the styles.

divyarajmasani committed:
- Uses CSS custom properties for some styling aspects for easy customization
- Adds base styles for lightbox and carousel view type.
- Adds styles for the stories block, uses pre-processor for the styles.

Commit:
Hide Readonly controls

Commit:
Run filter before outputting title

Co-authored-by: Jonny Harris <spacedmonkey@users.noreply.github.com>

Commit:
Nit pic: Render method call

Co-authored-by: Jonny Harris <spacedmonkey@users.noreply.github.com>

Commit:
Fix: Review changes

Commit:
Usage const instead of var

divyarajmasani committed:
Display story title for the stories on rearrange stories modal.

divyarajmasani committed:
- Fixes phpmd lint errors.

Commit:
Use admin_print_scripts hook to add data to page

divyarajmasani committed:
- Fixes phpunit tests.

divyarajmasani committed:
- Moves 'carousel-arrow' to new 'icons' directory.

divyarajmasani committed:
- Replaces 'load' event with '@wordpress/dom-ready' for lightbox script.
- Fixed some renderer code for redability.

divyarajmasani committed:
- Fix: Duplicate story meta was shown on select stories screen.

divyarajmasani committed:
- Moves SVGs to separte files and imports it, instead of hardcoding the SVGs in JS.

divyarajmasani committed:
- Moves classes for the story wrapper for better readablility.
- Updates to have archiveLink in backend block.

divyarajmasani committed:
- Adds basic support for RTL direction on Carousels.

divyarajmasani committed:
- Updates webpack config to build block styles for front-end and editor separately.

divyarajmasani committed:
- Updates few conditionals based on suggestions.
- Uses '@wordpress/hooks' package for the 'addFilter'.

divyarajmasani committed:
- Updates jest test for the Web Stories List block.

divyarajmasani committed:
- Cleans up CSS classes.
- Reduces repeatiton of the same style classes in markup.
- Updates phpunit tests to reflect the new classes.
- Fixes some known style issues.

divyarajmasani committed:
- Add specificity for the generic selectors for the block.

divyarajmasani committed:
Updates translator comment for the date shown.

divyarajmasani committed:
- Uses 'amp-img' for the AMP pages for the Web Stories List block story poster.

divyarajmasani committed:
- Adds 'since' for the methods and filter for phpdocs.

divyarajmasani committed:
- Uses localized 'isRTL' for the carousel.
- Updates carousel style for 'RTL' support.

Commit:
Move carousel icon to directory

Commit:
JS: Use const instead of var

Co-authored-by: Jonny Harris <spacedmonkey@users.noreply.github.com>

Commit:
ThemeSupport Trait

Commit:
JS: Use const instead of var

Co-authored-by: Jonny Harris <spacedmonkey@users.noreply.github.com>

Commit:
JS: Use const instead of var

Co-authored-by: Jonny Harris <spacedmonkey@users.noreply.github.com>

Commit:
JS: Use arrow function

Co-authored-by: Jonny Harris <spacedmonkey@users.noreply.github.com>

Commit:
JS: Use const instead of var

Co-authored-by: Jonny Harris <spacedmonkey@users.noreply.github.com>

Commit:
Escaping output

divyarajmasani committed:
- Reverts to use 'wp' hook to load 'load_amp_plugin_compat
- Fixes phpunit failed tests.

divyarajmasani committed:
- Removes workaround added to fix the issue of 'is_amp()' being not available for rest requests.

divyarajmasani committed:
- Updates carousel selector for modified markup.
- Fixes minor spacing issues for the carousel.

divyarajmasani committed:
- Updates block markup to reflect changes done in the final styles.

divyarajmasani committed:
- Fix jest tests for Web Stories List block.
- Fix eslint issue on carousel script.

divyarajmasani committed:
- Uses const for the max number of stories user can select for the select stories block.
- Adds a 'todo' to keep an eye on experimental APIs used.

divyarajmasani committed:
- Load Web Stories styles when rendering the block.

divyarajmasani committed:
- Updates configuration panel to use 'columnsCount' as prop for the styled component.
- Adds todo for removing 'field_states()' method from the block post (#189) gets merged.

divyarajmasani committed:
- Web Stories List Block: Update fieldState attribute to only store required data for the block.

Commit:
Fix: JS Lint

Commit:
Stories script class

Commit:
Fix: PHP Lint

Commit:
Rename: stories order function

Commit:
Remove IIFE in favour of webpack

Commit:
Restore: Loading of amp compatibility functions

divyarajmasani committed:
- Fix: updates components path used from 'dashboard'.
- Update: Removes unnecessary hook update in 'Plugin.php'.

Commit:
addEventListener for event capture

Commit:
Convert script data class to trait

Commit:
Remove script dependency

Commit:
Import domReady from package

Co-authored-by: Jonny Harris <spacedmonkey@users.noreply.github.com>

Commit:
JS: Use arrow function

Co-authored-by: Jonny Harris <spacedmonkey@users.noreply.github.com>

divyarajmasani committed:
- Selected Stories Block: Added config provider source to block directory.

divyarajmasani committed:
- Selected Stories Block: Uses localized object value for the storyAPI path.

divyarajmasani committed:
- Stories Block: Updates block constants, copied dependent constants from dashboard.

divyarajmasani committed:
- Embed Block: Fix e2e test block name for the embed stories.

divyarajmasani committed:
- Selected Stories: Addressed suggested changes for readability.

Commit:
TinyMCE class and ESLint rules

divyarajmasani committed:
Selected Stories Block:
- Removes 'wpAdapter' wrapper in-favour of direct use of 'apiFetch'.
- Removed unnecessary '_web_stories_envelope' from 'FetchSelectedStories' component.

Commit:
Remove lodash dependency

Commit:
Shorter syntax: String append

Co-authored-by: Pascal Birchler <pascal.birchler@gmail.com>

Commit:
Build Shortcode with template literals

Commit:
Fix: phpcs

Commit:
Fix: phpstan return type for method

Commit:
Fix: JS Doc

Co-authored-by: Pascal Birchler <pascal.birchler@gmail.com>

divyarajmasani committed:
Selected Stories Block:
- Uses 'addQueryArgs()' instead of 'queryString.stringifyUrl' for API providers.

Commit:
Fix: JS doc for reducer

Commit:
Update: Variable name to export

Commit:
Doc for patch

Commit:
Move prepareShortcode to utils

Commit:
Update doc comment

Co-authored-by: Jonny Harris <spacedmonkey@users.noreply.github.com>

Commit:
Check for function existence

Co-authored-by: Jonny Harris <spacedmonkey@users.noreply.github.com>

Commit:
Remove import of css

Commit:
Render markup outside of php

Commit:
Use Assets trait

Commit:
Add shortcode attr values in double quotes

divyarajmasani committed:
- Updates 'field_states()', removes unnecessary dependency on renderer.

divyarajmasani committed:
Web Stories Block:
- Adds support for shortcode to block transform.

divyarajmasani committed:
Web Stories Block:
- Experimental: Changes Block names to test with older post content.
- Embed block gets makeover as the new jumbo block.

divyarajmasani committed:
Web Stories Core Theme Support:
- Adds Customizer support for the core themes by default.

divyarajmasani committed:
Web Stories Block:
- Fixes undefined index notice in Renderer class when trying to use Customizer settings.
- Adds title show/hide toggle customizer control for 'circles' view.
- Replaces static options like 'order' and 'layouts' to use common utility functions.
- Fixes issue where multiple carousels were not working on the same page.
- Fixes issue for AMP where multiple lightbox of the same story were not working as expected.

divyarajmasani committed:
Web Stories Embed Block
- Moves embed block files to Stories List block directory.

Commit:
Add circle size field support

Commit:
Fix: PHPCS

Commit:
Fix: eslint issues

Commit:
Fix: Shadow rule reporting

divyarajmasani committed:
Web Stories Block:
- Updaets deprecation definition for Embed Block.

Commit:
Template tag function

Commit:
Fix: phpstan

Commit:
Fix: Eslint

divyarajmasani committed:
Deprecate Embed Block:
- Removes Embed Block in favour of Web Stories List.

divyarajmasani committed:
- Update deprecation definition to check for eligibility of the definition.

divyarajmasani committed:
- Fix e2e tests and some warnings for block components.

divyarajmasani committed:
- Fix: Select story modal was not closing if no stories were selected.
- Fix: Gutenberg warning for usage of custom block toolbar.
- Fix: Circles carousel were not initialized properly on mobile devices.
- Fix: Block styles were not loaded properly in editor.

divyarajmasani committed:
Reverts changes that has been moved to separate issue.

divyarajmasani committed:
Fix phpstan errors.

divyarajmasani committed:
- Updated Core Theme Support to accommodate the review.

divyarajmasani committed:
- Makes 'obj_id' of renderer class 'protected'.
- Adds 'data-id' for the web-stories containers.

divyarajmasani committed:
- Renders single lightbox wrapper element for all stories on the page.
- Updates lightbox css.
- Adds ligthboxes on 'wp_footer' action.

divyarajmasani committed:
Fix issue where web story were mis-matched in lightbox.

divyarajmasani committed:
Update failing tests.

divyarajmasani committed:
- Web Story Block renderer uses 'instance_id' instead of 'obj_id'.

divyarajmasani committed:
- Updates labels for block type switcher.
- Uses 'current_screen' hook to load tinyMCE.

divyarajmasani committed:
- Uses 'get_stylesheet' instead of 'get_template' for adding default support for Web Stories Customizer options.

divyarajmasani committed:
- Adds max-width to web stories inner wrapper.
- Fixes issue where archive link was not visible.
- Hides scrollbar for stories carousel.

Commit:
Use shorter syntax

Co-authored-by: Jonny Harris <spacedmonkey@users.noreply.github.com>

Commit:
Replace switch with if..else

Commit:
Remove radix

Commit:
Consistent wordings

Co-authored-by: Pascal Birchler <pascal.birchler@gmail.com>

Commit:
Change option name

Commit:
esc_attr for escaping

Co-authored-by: Jonny Harris <spacedmonkey@users.noreply.github.com>

Commit:
Separation of utility functions

Commit:
Customizer: Add sharp corner option

Commit:
Customizer: Remove story poster from grid

divyarajmasani committed:
- Removes static methods for rendering lightbox markup from 'Renderer.php'.
- Uses instance_id for the Renderer classes.

Commit:
Widget: Add archive label

Commit:
Widget: Number of columns field

Commit:
Widget: Sharp corners option

Commit:
Shortcode: Sharp corners

divyarajmasani committed:
- Removes 'default' from core themes support for customizer.

divyarajmasani committed:
- Removes unused method 'get_instance_id()' on Renderer.

Commit:
Shortcode: Archive label

divyarajmasani committed:
- Fixes 'z-index' to remove overlap.
- Adjusts positioning for the archive link.
- Adds bottom border for the stories render in the top.

divyarajmasani committed:
- Updates snapshot for the publishingFlow e2e test.

divyarajmasani committed:
- Adds styles for the core themes support.

divyarajmasani committed:
- Update block labes for consistency.

divyarajmasani committed:
- Loads theme specific styles only for web stories supported core themes.

Commit:
Fix: JS Tests

Commit:
Fix: Js Lint

divyarajmasani committed:
- Adds a wrapper div for stories rendered in the header.
- Uses 'path' to get the filename.

Commit:
Fix: Customizer controls

Commit:
Change shortcode tag

Commit:
Change archive label

Co-authored-by: Pascal Birchler <pascal.birchler@gmail.com>

divyarajmasani committed:
- Updates class used on body tag for Web Stories to 'has-web-stories'.

divyarajmasani committed:
- Removes 'assets()' loads style while rendering the stories.

divyarajmasani committed:
- Fixes string to replace all ocurrences of '-' for theme stylesheet slug.

divyarajmasani committed:
- Don't show excerpt for customizer stories by default.
- Check if readable file before enqueue for core theme styles.

divyarajmasani committed:
- Update block edit to use packages from '@web-stories-wp'

divyarajmasani committed:
- Resolves filter conflicts and potential maximum function call stack by replacing use of 'get_the_excerpt()' with '->post_excerpt'.
- Use template tag to render customizer stories.

divyarajmasani committed:
Stories block: Enable glider carousel for the editor screen.

divyarajmasani committed:
- Replaces use of '@web-stories-wp/date' with '@wordpress/date'.
- Updates eslintrc to resolve unwarranted flags in tinymce scripts.

divyarajmasani committed:
- Moves block related svg files to 'images/' to avoid SVGO overrides.
- Updates archive link styles for header stories.

divyarajmasani committed:
- Fix test for 'formatTime'.

divyarajmasani committed:
Fix: Publishing workflow e2e test.
Update: Time format token for 'am / pm'.

divyarajmasani committed:
- Handle unavaliable components on previous versions of WP.
- Updates to use alternate components when ToolbarGroup, Card components are not present.
- Mainly for version below WP v5.4.

divyarajmasani committed:
- Update configuration panel styles.

divyarajmasani committed:
- Fix: Carousel not initialised when adding block on new post.

divyarajmasani committed:
Merge pull request #323 from rtCamp/fix/package-updates

Fix/package updates

divyarajmasani committed:
Merge pull request #322 from rtCamp/feature/glider

Stories block: Enable glider carousel for the editor screen.

divyarajmasani committed:
Merge pull request #324 from rtCamp/fix/WP-5.3

Fix Web Stories block on WP 5.3

swissspidy commented on assets/src/web-stories-block/index.js:
```diff
@@ -28,9 +28,11 @@ import { registerBlockType } from '@wordpress/blocks';
 registerBlockType(name, settings);
 
-initializeTracking('Embed Block', false);
+initializeTracking('Web Stories List Block', false);
```
```suggestion
initializeTracking('Embed Block', false);
```

Don't want to skew our telemetry

swissspidy commented on assets/src/web-stories-block/block/deprecated.js:
```diff
@@ -97,11 +102,34 @@ save.propTypes = {
-];
+};
+
+const deprecated = [v1, v2];
```
```suggestion
// The deprecations in the array should be in reverse chronological order.
// This allows the block editor to attempt to apply the most recent and likely deprecations first, avoiding unnecessary and expensive processing.
const deprecated = [v2, v1];
```

swissspidy commented on assets/src/web-stories-block/block/transforms.js:
```diff
@@ -0,0 +1,139 @@
+    },
+    {
+      type: 'shortcode',
+      tag: 'stories',
```
```suggestion
      tag: 'web_stories',
```

divyarajmasani committed:
- Fix: Carousel layout was breaking in editor when try to reduce the number of stories.

divyarajmasani committed:
Merge pull request #325 from rtCamp/fix/broken-editor-carousel

Carousel layout was breaking in editor when try to reduce the number of stories.

spacedmonkey commented on includes/Block/Web_Stories_Block.php:
```diff
@@ -0,0 +1,351 @@
+	/**
+	 * Initializes the Web Stories embed block.
+	 *
+	 * @since 1.3.0
```
Do these since need to change?

spacedmonkey commented on includes/Model/Story.php:
```diff
@@ -36,12 +36,27 @@
+	 *
+	 * @var int
+	 */
+	protected $id;
```
Should id default to 0?

spacedmonkey commented on includes/Model/Story.php:
```diff
@@ -73,6 +88,20 @@ class Story {
+	 *
+	 * @var string
+	 */
+	protected $date;
```
Default to empty string?

spacedmonkey commented on includes/Model/Story.php:
```diff
@@ -73,6 +88,20 @@ class Story {
+	 *
+	 * @var string
+	 */
+	protected $author;
```
Default to empty string?

spacedmonkey commented on includes/Model/Story.php:
```diff
@@ -104,9 +133,10 @@ public function load_from_post( $_post ) {
-		$this->title  = get_the_title( $post );
-		$this->markup = $post->post_content;
-		$this->url    = (string) get_permalink( $post );
+		$this->title   = $post->post_title;
```
Why was this changed from using `get_the_title`?

swissspidy replied:
See explanation at https://github.com/rtCamp/web-stories-wp/pull/323#discussion_r581761395

Admittedly, we could still use `get_the_titlle()`. But `get_the_excerpt()` would cause an infinite loop.

spacedmonkey replied:
I understand `get_the_excerpt`, `get_the_title` I don't understand.

spacedmonkey commented on includes/Plugin.php:
```diff
@@ -207,6 +217,7 @@ class Plugin {
 	public function register() {
```
I would the opportunity to work on https://github.com/google/web-stories-wp/pull/5467 

This register method is overly long now. The problem with letting this go out, it means public context `$this->web_stories_block`

spacedmonkey commented on includes/Plugin.php:
```diff
@@ -250,13 +261,19 @@ public function register() {
 		$this->embed_shortcode = new Embed_Shortcode();
 		add_action( 'init', [ $this->embed_shortcode, 'init' ] );
 
+		$story_shortcode = new Stories_Shortcode();
```
Why isn't this a class property?

spacedmonkey commented on includes/Plugin.php:
```diff
@@ -306,6 +323,16 @@ public function register() {
+
+		add_action( 'widgets_init', [ $this, 'register_widgets' ] );
+
+		$tinymce = new TinyMCE();
```
Why isn't this a class property?

spacedmonkey commented on includes/Plugin.php:
```diff
@@ -250,13 +261,19 @@ public function register() {
 		// Gutenberg Blocks.
-		$this->embed_block = new Embed_Block();
-		add_action( 'init', [ $this->embed_block, 'init' ] );
+		$this->web_stories_block = new Web_Stories_Block();
```
There a chance of breakable here if were using this property to filter something in your plugin?

swissspidy replied:
I don't have a problem with that.

We've never declared these as extensible by others and [no plugin](https://wpdirectory.net/search/01EZS4CZW51ST8RM0F2PS0X7KT) is using this currently.

Don't need to bother with BC here.

spacedmonkey replied:
Agreed

spacedmonkey commented on includes/TinyMCE.php:
```diff
@@ -0,0 +1,121 @@
+<?php
```
Should we create a new namespace called Admin, add TinyMCE and Customizer ( and Admin ) to it?

spacedmonkey commented on includes/Traits/Stories_Script_Data.php:
```diff
@@ -0,0 +1,108 @@
+
+		$data = [
+			'orderlist' => $order_list,
+			'icon'      => trailingslashit( WEBSTORIES_ASSETS_URL ) . 'images/widget/carousel-icon.png',
```
Should this be a svg instead?

spacedmonkey commented on includes/functions.php:
```diff
@@ -0,0 +1,196 @@
+use Google\Web_Stories\Stories_Renderer\FieldState\CarouselView;
+use Google\Web_Stories\Stories_Renderer\FieldState\CircleView;
+use Google\Web_Stories\Stories_Renderer\FieldState\ListView;
+
```
Sinces need to be added here.

spacedmonkey commented on tests/phpunit/tests/Plugin.php:
```diff
@@ -36,7 +36,7 @@ public function test_register() {
 		$this->assertSame( 10, has_action( 'admin_init', [ $plugin->admin, 'init' ] ) );
 		$this->assertSame( 5, has_action( 'admin_init', [ $plugin->database_upgrader, 'init' ] ) );
-		$this->assertSame( 10, has_action( 'init', [ $plugin->embed_block, 'init' ] ) );
+		$this->assertSame( 10, has_action( 'init', [ $plugin->web_stories_block, 'init' ] ) );
```
We should add all the new class properties here.

spacedmonkey committed:
Merge branch 'main' into feature/embed-stories

spacedmonkey commented on assets/src/carousel/index.js:
```diff
@@ -0,0 +1,185 @@
+/*
```
This should be it's own package. It would be extremely simple to move it.

spacedmonkey replied:
Fixed.

spacedmonkey commented on assets/src/lightbox/index.js:
```diff
@@ -0,0 +1,29 @@
+/*
```
This should be it's own package. It would be extremely simple to move it.

spacedmonkey commented on assets/src/lightbox/web-stories-lightbox.js:
```diff
@@ -0,0 +1,91 @@
+ *
+ * Displays the story that user interacted with in a lightbox.
+ */
+class Lightbox {
```
Is there a reason we are using a class here? We are using functional programming everywhere else.

swissspidy replied:
I think we've discussed this before. For this use case it seems quite fitting and easier to read & maintain this way.

spacedmonkey commented on assets/src/tinymce/__tests__/components/Modal.test.js:
```diff
@@ -0,0 +1,93 @@
+/*
```
The directory should be renamed `tests` from `__tests__`

ankitrox replied:
@spacedmonkey This was discussed before [here](https://github.com/rtCamp/web-stories-wp/pull/194#discussion_r565848120)

spacedmonkey replied:
Fixed.

spacedmonkey commented on assets/src/tinymce/web-stories.js:
```diff
@@ -0,0 +1,77 @@
+/*
```
This too should be its own package. It will make managing dependencies much easier.

spacedmonkey replied:
Fixed.

spacedmonkey commented on assets/src/web-stories-block/block/api/useStoryApi.js:
```diff
@@ -0,0 +1,144 @@
+/**
+ * External dependencies
+ */
+import { useCallback, useMemo, useReducer } from 'react';
```
These should be loaded from `@wordpress/elements`

spacedmonkey commented on assets/src/web-stories-block/block/api/useStoryApi.js:
```diff
@@ -0,0 +1,144 @@
+  STORIES_PER_REQUEST,
+  STORY_STATUS,
+} from '../constants';
+import storyReducer, {
```
I would like to add a todo here to note the interdepencies and make sure we get rid of this at some point.

spacedmonkey commented on assets/src/web-stories-block/block/api/useUserApi.js:
```diff
@@ -0,0 +1,65 @@
+/**
+ * External dependencies
+ */
+import { useCallback, useMemo, useState } from 'react';
```
Use from `@wordpress/elements`.

spacedmonkey commented on assets/src/web-stories-block/block/block-types/selected-stories/components/cardGridItem.js:
```diff
@@ -0,0 +1,44 @@
+/**
+ * Internal dependencies
+ */
+import { CardGrid, CardGridItem } from '../../../../../dashboard/components';
```
Add todo to note the interdepencies

spacedmonkey commented on assets/src/web-stories-block/block/block-types/selected-stories/components/itemOverlay.js:
```diff
@@ -0,0 +1,146 @@
+/**
+ * Internal dependencies
+ */
+import { PageSizePropType } from '../../../../../dashboard/types';
```
Add todo to note the interdepencies

spacedmonkey commented on assets/src/web-stories-block/block/block-types/selected-stories/edit.js:
```diff
@@ -0,0 +1,148 @@
+/**
+ * Internal dependencies
+ */
+import theme from '../../../../dashboard/theme';
```
Add todo to note the interdepencies

spacedmonkey commented on assets/src/web-stories-block/block/block-types/selected-stories/selectStories.js:
```diff
@@ -0,0 +1,296 @@
+/**
+ * Internal dependencies
+ */
+import { PageSizePropType } from '../../../../dashboard/types';
```
Add todo to note the interdepencies

spacedmonkey commented on assets/src/web-stories-block/block/block-types/selected-stories/sortStories.js:
```diff
@@ -0,0 +1,182 @@
+/**
+ * Internal dependencies
+ */
+import { PageSizePropType } from '../../../../dashboard/types';
```
Add todo to note the interdepencies

spacedmonkey commented on assets/src/web-stories-block/block/block-types/selected-stories/storyPicker.js:
```diff
@@ -0,0 +1,341 @@
+  display: flex;
+  justify-content: space-between;
+  align-items: center;
+  background: #f3f3f3;
```
There are lots of hardcoded colors here. Are should these values be in a theme somewhere.

spacedmonkey commented on assets/src/web-stories-block/block/block-types/selected-stories/storyPreview.js:
```diff
@@ -0,0 +1,102 @@
+/**
+ * Internal dependencies
+ */
+import {
+  CardPreviewContainer,
+  CardTitle,
+} from '../../../../dashboard/components';
+import { StoryPropType, PageSizePropType } from '../../../../dashboard/types';
```
Add todo to note the interdepencies

spacedmonkey commented on assets/src/web-stories-block/block/components/storiesBlockConfigurationPanel.js:
```diff
@@ -0,0 +1,137 @@
+const TypeMedia = styled(CardMedia)`
+  display: flex;
+  align-items: center;
+  color: #347bb5;
```
Hardcoded color

spacedmonkey commented on assets/src/web-stories-block/block/components/typeaheadAuthorSearch.js:
```diff
@@ -0,0 +1,71 @@
+/**
+ * Internal dependencies
+ */
+import { Typeahead } from '../../../dashboard/components';
```
Add todo to note the interdepencies

spacedmonkey commented on assets/src/web-stories-block/block/index.js:
```diff
@@ -22,44 +22,49 @@ import { __ } from '@wordpress/i18n';
+  description: __('Embed Web Stories.', 'web-stories'),
   category,
   icon,
   keywords: [
```
For some reason, I thought that keywords are limited at 3. But I am not sure where I got that from. Latest and selected feel like a little generic as keywords.

swissspidy commented on assets/images/web-stories/block/carousel-arrow.svg:
```diff
@@ -0,0 +1,3 @@
+<svg width="9" height="14" viewBox="0 0 9 14" fill="none" xmlns="http://www.w3.org/2000/svg">
```
I don't see why this should be under `assets/images`

spacedmonkey commented on includes/Widgets/Stories.php:
```diff
@@ -0,0 +1,476 @@
+				'classname'     => 'widefat archive_label stories-widget-field',
+				'wrapper_class' => 'archive_label_wrapper',
+				'value'         => $number_columns,
+				'label_before'  => true,
```
We should add some validation here, so min of 0, a reasonable max.

spacedmonkey commented on includes/Widgets/Stories.php:
```diff
@@ -0,0 +1,476 @@
+				'classname'     => 'widefat number-stories stories-widget-field',
+				'wrapper_class' => 'number-stories_wrapper',
+				'value'         => $number,
+				'label_before'  => true,
```
There should be min and max attrs on number input

spacedmonkey commented on includes/Widgets/Stories.php:
```diff
@@ -0,0 +1,476 @@
+		$instance['show_date']         = ( isset( $new_instance['show_date'] ) ) ? 1 : '';
+		$instance['archive_link']      = ( isset( $new_instance['archive_link'] ) ) ? 1 : '';
+		$instance['image_align_right'] = ( isset( $new_instance['image_align_right'] ) ) ? 1 : '';
+		$instance['number']            = min( absint( $new_instance['number'] ), 20 );
```
```suggestion
		$instance['number']            = min( 1, absint( $new_instance['number'] ), 20 );
```

spacedmonkey commented on includes/Widgets/Stories.php:
```diff
@@ -0,0 +1,476 @@
+		$instance['archive_link']      = ( isset( $new_instance['archive_link'] ) ) ? 1 : '';
+		$instance['image_align_right'] = ( isset( $new_instance['image_align_right'] ) ) ? 1 : '';
+		$instance['number']            = min( absint( $new_instance['number'] ), 20 );
+		$instance['circle_size']       = min( absint( $new_instance['circle_size'] ), 200 );
```
```suggestion
		$instance['circle_size']       = max( min( absint( $new_instance['circle_size'] ), 80 ), 200 );
```

spacedmonkey committed:
Fix quantize issue.

Commit:
Fix: View type support

divyarajmasani committed:
- Fix: Stories are overlapped by feature images and some other elements on pages for AMP requests.
- Fix: Archive link is not visible on twenty-sixteen theme.

Commit:
Merge pull request #338 from rtCamp/fix/view-type-support

Fix: View type support

swissspidy commented on includes/Traits/ThemeSupport.php:
```diff
@@ -0,0 +1,76 @@
+		$theme_support['grid-columns-default'] = (int) $theme_support['grid-columns-default'];
+
+		return $theme_support;
+	}
```
This is an API part I want to change before release.

Having fields like `xyz` and `xyz-default` is not very intuitive.

I'd rather have something like

```php
'excerpt' => [
  'enabled' => true,
  'default' => false
]
```

At  the same time it's confusing that `circle-size-default` has no `circle-size` counterpart and that it's a top level item, even though it's specific to the circle layout.

Could do

```php
'circle_sizee' => [
  'default' => 150
]
```

Or perhaps een

```php
'view_types' => [
  'circles' => [
    'label' => __( 'Circle Carousel', 'web-stories' ), // why need to provide this though?
    'options' => [
      'circle_size' => [
        'default' => 150
      ]
    ]
  ]
]
```

swissspidy commented on includes/functions.php:
```diff
@@ -0,0 +1,196 @@
+ *
+ * @return GridView|ListView|CircleView|CarouselView
+ */
+function get_field( $view = 'grid' ) {
```
This shouldn't be a function in this file.

Let's move it to a factory class or something.

spacedmonkey replied:
I agree, this feels weird here.

swissspidy commented on includes/functions.php:
```diff
@@ -0,0 +1,196 @@
+ *
+ * @return array
+ */
+function fields_states() {
```
This shouldn't be a function in this file.

Let's move it to a factory class or something.

swissspidy commented on includes/functions.php:
```diff
@@ -0,0 +1,196 @@
+ *
+ * @return array
+ */
+function get_stories_order() {
```
Don't see why this is a  separate function in this file. Should probably be moved elsewhere.

spacedmonkey replied:
I think this should live in the ThemeSupport trait as a method.

swissspidy commented on includes/functions.php:
```diff
@@ -0,0 +1,196 @@
+ *
+ * @return mixed|void
+ */
+function get_layouts() {
```
Don't see why this is a  separate function in this file. Should probably be moved elsewhere.

spacedmonkey replied:
Maybe this should be a trait, it is used in a number of places.

swissspidy commented on includes/functions.php:
```diff
@@ -0,0 +1,196 @@
+			'oldest'               => __( 'Oldest', 'web-stories' ),
+			'alphabetical'         => __( 'A -> Z', 'web-stories' ),
+			'reverse-alphabetical' => __( 'Z -> A', 'web-stories' ),
+		]
```
I've always found this very confusing, it mixes order and orderby in one list.

We should instead split this up under the hood.

So `alphabetical` becomes `orderby: title, order: ASC`

swissspidy commented on includes/Stories_Renderer/Renderer.php:
```diff
@@ -0,0 +1,746 @@
+	public function is_amp_request() {
+
+		return is_amp();
+	}
```
This `is_amp_request` method is unnecessary. Can just use `is_amp` directly.

swissspidy commented on includes/Stories_Renderer/Renderer.php:
```diff
@@ -0,0 +1,746 @@
+				>
+				</amp-img>
+			<?php } else { ?>
+				<img src="<?php echo esc_url( $poster_url ); ?>" alt="<?php echo esc_attr( $story_data->get_title() ); ?>">
```
Need `width` and `height` attribute here

divyarajmasani committed:
Merge pull request #339 from rtCamp/fix/core-theme-tweaks

Core Theme style updates.

divyarajmasani committed:
Merge remote-tracking branch 'upstream/feature/embed-stories' into feature/embed-stories

spacedmonkey commented on includes/Customizer.php:
```diff
@@ -0,0 +1,568 @@
+					return $this->is_option_enabled( 'show_stories' );
+				},
+			]
+		);
```
These should be broken into two fields as noted elsewhere. Order or order by. Having this as one field feels wrong. You agree @swissspidy ?

spacedmonkey commented on includes/Integrations/Core_Themes_Support.php:
```diff
@@ -0,0 +1,133 @@
+	public function embed_web_stories() {
+		if ( is_readable( WEBSTORIES_PLUGIN_DIR_PATH . '/assets/css/web-stories-theme-style-' . get_stylesheet() . '.css' ) ) {
+			$this->enqueue_style( 'web-stories-theme-style-' . get_stylesheet(), [ Renderer::STYLE_HANDLE ] );
+		}
```
```suggestion
		$stylesheet = get_stylesheet();
		if ( is_readable( sprintf( '%s/assets/css/web-stories-theme-style-%s.css', WEBSTORIES_PLUGIN_DIR_PATH, $stylesheet ) ) {
			$this->enqueue_style( 'web-stories-theme-style-' . $stylesheet, [ Renderer::STYLE_HANDLE ] );
		}
```

This feels a little cleaner here.

spacedmonkey replied:
Fixed in https://github.com/google/web-stories-wp/pull/6543/commits/945b8428cd8d007795dd1a81360251ed4599e1f2

spacedmonkey commented on includes/Story_Query.php:
```diff
@@ -0,0 +1,180 @@
+
+		$query_args    = $this->get_query_args();
+		$stories_query = new WP_Query( $query_args );
+		$posts         = ( ! empty( $stories_query->posts ) && is_array( $stories_query->posts ) ) ? $stories_query->posts : [];
```
```suggestion
		$stories_query = new WP_Query();
		$posts         = $storires_query->query( $query_args );
```

A little cleaner.

spacedmonkey committed:
Make Carousel into a package. (#6575)

* Make Carousel into a package.

* Change package name

spacedmonkey committed:
Make lightbox in a package (#6576)

* Make lightbox into a package.

* Fix lock file.

spacedmonkey committed:
Make widget into a package. (#6577)

spacedmonkey committed:
Fix since versions.

spacedmonkey committed:
Make tinymce into a package. (#6578)

* Make tinymce into a package.

* Feedback.

* Fixes.

* Remove unneeded react

* Buttons to button

* Change lock file.

* Change main file.

* Change path.

* change path again.

spacedmonkey committed:
Change copyright year.

spacedmonkey committed:
Change is_amp to trait. (#6583)

* Change to trait.

* Use is_amp.

* Make it public.

swissspidy commented on assets/src/web-stories-block/block/transforms.js:
```diff
@@ -0,0 +1,139 @@
+              show_date: 'true' === date ? true : false,
+              show_excerpt: 'true' === excerpt ? true : false,
+              show_image_align: 'true' === image_align ? true : false,
+              show_title: 'true' === title ? true : false,
```
```suggestion
              show_archive_link: 'true' === archive_link,
              show_author: 'true' === author,
              show_date: 'true' === date,
              show_excerpt: 'true' === excerpt,
              show_image_align: 'true' === image_align,
              show_title: 'true' === title,
```

swissspidy committed:
Merge branch 'main' into feature/embed-stories

swissspidy commented on assets/src/web-stories-block/block/components/storyCard.js:
```diff
@@ -0,0 +1,105 @@
+  return (
+    <div className={singleStoryClasses}>
+      <div className="web-stories-list__story-poster">
+        <img src={poster} alt={title} />
```
Should add width/height and only display if there is a poster.

spacedmonkey commented on packages/tinymce-button/src/components/Modal.js:
```diff
@@ -0,0 +1,212 @@
+                dispatch(name).toggleModal(false);
+              }}
+            >
+              {__('Okay', 'web-stories')}
```
```suggestion
              {__('Insert', 'web-stories')}
```

spacedmonkey replied:
@swissspidy Are you happy with this change? What different wording?

swissspidy replied:
Addressing this in #6597

spacedmonkey committed:
Cleaner code use of WP_Query

spacedmonkey committed:
Refactor stylesheet lookup.

spacedmonkey committed:
Fix lint.

swissspidy committed:
Fix various nits in #6543 (#6589)

swissspidy committed:
Improve consistency between various embed methods (#6597)

Co-authored-by: Jonny Harris <spacedmonkey@users.noreply.github.com>

divyarajmasani committed:
[Block] Minor issues (#6626)

* [Block]
- Fix: author filter
- Fix: Stories and Author dropdowns were not shown for selected stories modal.
- Fix: undefined index for 'sharp_corners' and 'image_alignment'.

* - Fix phpunit test for Stories widget.

swissspidy committed:
Embeds: Use placeholder when poster image is missing (#6625)

spacedmonkey committed:
Add some basic widget validation. (#6627)

swissspidy committed:
Add custom icon to stories widget (#6628)

swissspidy committed:
Merge branch 'main' into feature/embed-stories

spacedmonkey commented on includes/Block/Web_Stories_Block.php:
```diff
@@ -0,0 +1,346 @@
+	/**
+	 * Initializes the Web Stories embed block.
+	 *
+	 * @since 1.3.0
```
```suggestion
	 * @since 1.5.0
```

spacedmonkey reviewed (approved):
There is still some way to get with this functionality. Some restructuring, improve PHPUnit & E2E tests and refactor. But I think it is good enough to ship. Really excited to see this merged 🎉

swissspidy committed:
Merge branch 'main' into feature/embed-stories

swissspidy committed:
Update `since` tags

swissspidy committed:
Shortcode: add missing orderby

swissspidy committed:
Fix shortcode attributes

divyarajmasani committed:
Stop scrolling when lightbox is open (#6660)

swissspidy committed:
Merge branch 'main' into feature/embed-stories

swissspidy committed:
Fix show_archive_link shortcode attribute

swissspidy committed:
Fix selected stories block crashing

BrittanyIRL committed:
decrease button and pill transition time from 0.6s to 0.3s in design system (#6668)

BrittanyIRL committed:
Design System: Firefox bugfix for context menus (#6670)

* this useEffect is firing before the animated context menu is usable (`isReady`) state. the isReady state is not available at this level of the menu. Chrome is able to figure this out and works fine, however firefox doesnt - so menu items that are buttons dont ever trigger an onClick. It seems like this was just a nice enhancement from the existing functionality, not something that is key to the menu working. As such, I opted to remove the useEffect hook ahead of release since I couldnt figure out an alternative in a timely manner.

* Fix accessibility after animating

* Rever my change it does not work

* karma catch up for context menu keyboard behavior

Co-authored-by: samwhale <samuelestrella226@gmail.com>

spacedmonkey committed:
ESLint: Enable react/jsx-boolean-value rule (#6677)

swissspidy committed:
Lint fixes

spacedmonkey committed:
Widget: PHP Tests (#6682)

Co-authored-by: Pascal Birchler <pascalb@google.com>

swissspidy committed:
WordPress 5.7 Compatibility (#6674)

Co-authored-by: Pascal Birchler <pascalb@google.com>

swissspidy committed:
Merge branch 'main' into feature/embed-stories

spacedmonkey committed:
Widget notice errors (#6711)

spacedmonkey committed:
TinyMCE Button: PHP Tests (#6732)

spacedmonkey committed:
Infrastructure: Move embed.css to the other CSS files (#6736)

spacedmonkey committed:
Widget: E2E Tests (#6709)

spacedmonkey committed:
Classic Editor: E2E tests for TinyMCE button (#6705)

swissspidy committed:
Merge branch 'main' into feature/embed-stories

spacedmonkey committed:
Update package-lock.json file.

swissspidy committed:
Merge branch 'main' into feature/embed-stories

swissspidy committed:
Improve block styling when displaying archive links

swissspidy committed:
hot fixes for glider-js

swissspidy committed:
fix lock file

spacedmonkey committed:
Merge branch 'main' into feature/embed-stories

swissspidy committed:
Prevent notices

swissspidy committed:
Merge branch 'main' into feature/embed-stories

swissspidy committed:
Normalize composer.json

swissspidy committed:
fix tinymce script

divyarajmasani committed:
Rewind story in the lightbox (#6771)

- Feature: rewind story upon closing the lightbox.

swissspidy committed:
Merge branch 'main' into feature/embed-stories

swissspidy committed:
Merge branch 'main' into feature/embed-stories

swissspidy committed:
Make customizer e2e test more robust

swissspidy committed:
Lint fix

swissspidy committed:
Fix WP-CLI command

swissspidy committed:
Fix selector name

swissspidy committed:
Make tinymce button e2e test more robust

swissspidy committed:
Remove debug cruft

swissspidy committed:
Fix widget test on WP 5.3

swissspidy committed:
Merge branch 'main' into feature/embed-stories

swissspidy committed:
Merge branch 'main' into feature/embed-stories

swissspidy committed:
Update widget e2e test
